plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -83,24 +83,17 @@ impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
                 );
             } else {
                 // The rest leaf layers
-                let new_leaves: Vec<Vec<F>> = cap
-                    .iter()
-                    .enumerate()
-                    .map(|(i, cap_hash)| {
-                        let mut new_hash = Vec::with_capacity(NUM_HASH_OUT_ELTS + cur[i].len());
-                        new_hash.extend(&cap_hash.to_vec());
-                        new_hash.extend(&cur[i]);
-                        new_hash
-                    })
-                    .collect();
-                cap.clear();
-                cap.reserve_exact(next_cap_len);
+                let new_leaves = cap;
+                cap = Vec::with_capacity(next_cap_len);
                 let tmp_cap_buf = capacity_up_to_mut(&mut cap, next_cap_len);
-                fill_digests_buf::<F, H>(
+                fill_digests_buf_custom::<F, H, _, _>(
                     &mut digests_buf[digests_buf_pos..(digests_buf_pos + num_tmp_digests)],
                     tmp_cap_buf,
                     &new_leaves[..],
                     next_cap_height,
+                    |i, cap_hash| {
+                        H::hash_or_noop_iter(chain!(cap_hash.into_iter(), cur[i].iter().copied()))
```

Comment:
This smells like it wants to be a `zip` or so?

Reply:
`zip`? Why? The goal is to concatenate them before hashing.